File: ios/LowerBackStretching/Notifications/ReminderController.swift
import Foundation

/// Single source of truth for "user changed the reminder". Persists the
/// preference to UserDefaults (the same store @AppStorage reads/writes) and
/// re-(scheduling|cancelling) the UNCalendarNotificationTrigger so the two
/// never drift. Called by Settings and Onboarding.
enum ReminderController {
    /// `defaults` is injectable for tests; production code uses `.standard`
    /// which is what `@AppStorage` reads/writes.
    static func apply(
        enabled: Bool,
        hour: Int,
        minute: Int,
        defaults: UserDefaults = .standard
    ) {
        defaults.set(enabled, forKey: ReminderDefaults.enabledKey)
        defaults.set(hour,    forKey: ReminderDefaults.hourKey)
        defaults.set(minute,  forKey: ReminderDefaults.minuteKey)

        if enabled {
            ReminderManager.schedule(hour: hour, minute: minute)
        } else {
            ReminderManager.cancel()
        }
    }
}

File: android/app/src/main/java/com/lowerbackstretching/notifications/ReminderController.kt
package com.lowerbackstretching.notifications

import android.content.Context
import com.lowerbackstretching.data.Prefs

/**
 * Single source of truth for "user changed the reminder". Persists the
 * preference and re-(scheduling|cancelling) the AlarmManager alarm so the
 * two never drift. Called by Settings and Onboarding.
 */
suspend fun Prefs.applyReminder(context: Context, enabled: Boolean, hour: Int, minute: Int) {
    setReminder(enabled, hour, minute)
    if (enabled) {
        ReminderScheduler.schedule(context, hour, minute)
    } else {
        ReminderScheduler.cancel(context)
    }
}

/**
 * Single source of truth for "user changed the streak-at-risk nudge."
 * Persists the preference and re-(scheduling|cancelling) the AlarmManager
 * alarm so the two never drift. The receiver gates posting per-day on
 * the current streak and whether the user already stretched today.
 */
suspend fun Prefs.applyStreakNudge(context: Context, enabled: Boolean) {
    setStreakNudgeEnabled(enabled)
    if (enabled) {
        ReminderScheduler.scheduleStreakNudge(context)
    } else {
        ReminderScheduler.cancelStreakNudge(context)
    }
}
